components/places/src/storage/bookmarks/public.rs

```diff
@@ -0,0 +1,582 @@
+///
+/// It also produces the protobuf message type directly, rather than
+/// add a special variant of this bookmark type just for this function.
+pub fn fetch_bookmark(
+    db: &PlacesDb,
+    item_guid: &SyncGuid,
+    get_direct_children: bool,
+) -> Result<Option<PublicNode>> {
+    let _tx = db.unchecked_transaction()?;
+    let scope = db.begin_interrupt_scope();
+    let bookmark = fetch_bookmark_in_tx(db, item_guid, get_direct_children, &scope)?;
+    // Note: We let _tx drop (which means it does a rollback) since it doesn't
+    // matter, we just are using a transaction to ensure things don't change out
+    // from under us, since this executes more than one query.
+    Ok(bookmark)
+}
+
+fn get_child_guids(db: &PlacesDb, parent: RowId) -> Result<Vec<SyncGuid>> {
+    Ok(db.query_rows_into(
+        "SELECT guid FROM moz_bookmarks
+         WHERE parent = :parent
+         ORDER BY position ASC",
+        &[(":parent", &parent)],
+        |row| row.get_checked(0),
+    )?)
+}
+
+fn fetch_bookmark_child_info(
+    db: &PlacesDb,
+    parent: &RawBookmark,
+    get_direct_children: bool,
+    scope: &crate::db::InterruptScope,
+) -> Result<(Option<Vec<SyncGuid>>, Option<Vec<PublicNode>>)> {
+    if parent.bookmark_type != BookmarkType::Folder {
+        return Ok((None, None));
+    }
+    // If we already know that we have no children, don't
+    // bother querying for them.
+    if parent.child_count == 0 {
+        return Ok(if get_direct_children {
+            (None, Some(vec![]))
+        } else {
+            (Some(vec![]), None)
+        });
+    }
+    if !get_direct_children {
+        // Just get the guids.
+        return Ok((Some(get_child_guids(db, parent.row_id)?), None));
+    }
+    // Fetch children. the future this should probably be done by allowing a
+    // depth parameter to be passed into fetch_tree.
+    let child_nodes = get_raw_bookmarks_with_parent(db, parent.row_id)?
+        .into_iter()
+        .map(|kid| {
+            let child_guids = if kid.bookmark_type == BookmarkType::Folder {
+                if kid.child_count == 0 {
+                    Some(vec![])
+                } else {
+                    Some(get_child_guids(db, kid.row_id)?)
+                }
+            } else {
+                None
+            };
+            scope.err_if_interrupted()?;
+            Ok(PublicNode::from(kid).with_children(child_guids, None))
+        })
+        .collect::<Result<Vec<_>>>()?;
+    Ok((None, Some(child_nodes)))
+}
+
+// Implementation of fetch_bookmark
+fn fetch_bookmark_in_tx(
+    db: &PlacesDb,
+    item_guid: &SyncGuid,
+    get_direct_children: bool,
+    scope: &crate::db::InterruptScope,
+) -> Result<Option<PublicNode>> {
+    // get_raw_bookmark doesn't work for the bookmark root, so we just return None explicitly
+    // (rather than erroring). This isn't ideal, but there's no point to fetching the "true"
+    // bookmark root without fetching its children too, so whatever.
```

Comment:
Hmm, I think you said on Slack you wanted to fix `get_raw_bookmark` not to fail in that case...decided not to?

Reply:
Oh, stale comment.